### PR TITLE
REST API for applications

### PR DESCRIPTION
**Pull request by algorandskiy**
* Add `v2/accounts/creator/{id}` endpoint [as discussed ](https://github.com/algorand/go-algorand/issues/1141)
* Implement rest client methods and `libgoal` wrappers
* Use new libgoal functions in `goal app`
* Remove `v1/applications`, `v1/application`
* Leave applications support in v1 transactions

### CODE REVIEW

**Comment by algorandskiy** [comment]
@winder @algobolson

File: daemon/algod/api/server/v2/handlers.go [hunk 1]
```diff
@@ -92,24 +92,64 @@ func (v2 *Handlers) AccountInformation(ctx echo.Context, address string, params
...
+		return internalError(ctx, err, errInternalFailure, v2.Log)
+	}
+
+	response := generated.AccountResponse(account)
+	return ctx.JSON(http.StatusOK, response)
+}
+
+// AccountInformationCreatorOf gets account information for a creator of an asset or application
+// (GET /v2/accounts/creator/{ident})
+func (v2 *Handlers) AccountInformationCreatorOf(ctx echo.Context, ident uint64, params generated.AccountInformationCreatorOfParams) error {
```
Comment (winder) [inline daemon/algod/api/server/v2/handlers.go:107]: Unless I'm mistaken, this function can be greatly simplified by reusing `AccountInformation`:
```
func (v2 *Handlers) AccountInformationCreatorOf(ctx echo.Context, ident uint64, params generated.AccountInformationCreatorOfParams) error {
    ledger := v2.Node.Ledger()
    aidx := basics.AssetIndex(ident)
    creator, ok, err := ledger.GetAssetCreator(aidx)
    if err != nil {
        return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
    }
    return v2.AccountInformation(ctx, creator, generated.AccountInformationParams{ format: params.format })
}
```

**Comment by brianolson** [comment]
LGTM, I'm a little surprised that list-applications is going away, but I guess it's not crucial.

File: daemon/algod/api/server/v2/account.go [hunk 2]
```diff
@@ -334,3 +335,23 @@ func ApplicationParamsToAppParams(gap *generated.ApplicationParams) basics.AppPa
 
 	return ap
 }
+
+func getAssetCreatorsForHoldings(ledger *data.Ledger, holdings map[basics.AssetIndex]basics.AssetHolding) map[basics.AssetIndex]string {
+	assetsCreators := make(map[basics.AssetIndex]string, len(holdings))
+	if len(holdings) > 0 {
```
Comment (algomaxj) [inline daemon/algod/api/server/v2/account.go:341]: unnecessary check

File: daemon/algod/api/algod.oas2.json [hunk 3]
```diff
@@ -229,6 +229,74 @@
         }
       ]
     },
+    "/v2/accounts/creator/{ident}": {
```
[the remark this answers is not in the document]
Comment (winder) [reply daemon/algod/api/algod.oas2.json:232]: We shouldn't merge applications/assets, that's an implementation detail that we should avoid adding to the API semantics.

Rotem had some concerns with the endpoint naming as well, so please follow up with him before committing anything.

**Comment by algomaxj** [comment]
Looks good to me once feedback is addressed.

**Comment by algorandskiy** [comment]
> Rotem had some concerns with the endpoint naming as well, so please follow up with him before committing anything.

@winder @algorotem
I'm happy with two endpoints or whatever you want. Please make a call.

File: daemon/algod/api/client/restClient.go [hunk 4]
```diff
@@ -349,14 +343,21 @@ func (blob *Blob) SetBytes(b []byte) {
 	*blob = b
 }
 
-// RawAccountInformationV2 gets the raw AccountData associated with the passed address
-func (client RestClient) RawAccountInformationV2(address string) (response []byte, err error) {
+// AccountInformationRawV2 gets the raw AccountData associated with the passed address
+func (client RestClient) AccountInformationRawV2(address string) (response []byte, err error) {
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/api/client/restClient.go:347]: No, this is `Raw` function. there is another one
```
func (client RestClient) AccountInformationV2(address string) (response generatedV2.Account, err error)
```
that returns decoded account

File: daemon/algod/api/client/restClient.go [hunk 5]
```diff
@@ -349,14 +343,21 @@ func (blob *Blob) SetBytes(b []byte) {
...
 	var blob Blob
 	err = client.getRaw(&blob, fmt.Sprintf("/v2/accounts/%s", address), rawAccountParams{Format: "msgpack"})
 	response = blob
 	return
 }
 
+// AccountInformationCreatorOfRaw gets the basics.BalanceRecord associated
+// with the passed asset or application index
+func (client RestClient) AccountInformationCreatorOfRaw(index uint64) (response []byte, err error) {
+	err = client.get(&response, fmt.Sprintf("/v2/accounts/creator/%d", index), rawAccountParams{Format: "msgpack"})
```
[the remark this answers is not in the document]
Comment (algorandskiy) [reply daemon/algod/api/client/restClient.go:357]: Same story

**Comment by algorotem** [comment]
We already have an account information endpoint, there is no need to introduce a new one. 
Similarly to what we did in v1 for assets, we should simply add an `applications` field to `v2/accounts/{addr}`. if the user happened to be a creator of an app, the information will show up under the user's account data. 

If there is a need to pull on a specific application info, we should introduce a `v2/applications/{AppID}` endpoint.

**Comment by algorandskiy** [comment]
@algorotem 
Add `asset` field to where? I need to get account info by application or asset id without knowing creator address. How would `v2/accounts/{addr}` endpoint look like in this case?

**Comment by algorotem** [comment]
Sorry, now I see that I misunderstood the meaning of `{ident}`. In this case, I suggest creating `v2/applications/{id}` which will return the application info including the `creator`. Then you can simply access the info in `/v2/accounts/{addr}`.
Same for `assets`.

**Comment by algorandskiy** [comment]
All right. I'll open another PR instead since it will be another implementation.

**Comment by algorandskiy** [comment]
Replaced by https://github.com/justicz/go-algorand/pull/138

**Pull request closed without merging**